posts/frangeris/01.reusable-dynamic-modal-on-vue3.md

```diff
@@ -0,0 +1,215 @@
+  const { isOpen, view, actions } = storeToRefs(modal);
+</script>
+```
+
+> `<component>` is the way that `vue` handle dynamic components where `:is` receive the view that needs to be rendered, more info [here](https://vuejs.org/api/built-in-special-elements.html#component), it also support using the `v-model` directive to be able to share values dynamically, but how does this works?
+
+Once we have the view mounted inside `<component>` we can make use of `v-model` to send and update reactive references emitting events from the view we are mounting, internally from the view we want to render: `emit` a `"update:modelValue"` event that will update the reference passed via `v-model` back, eg:
+
+```html
+<!-- MyViewToRender.vue -->
+<script lang="ts" setup>
+  import { watch } from "vue";
+
+  // no need to import defineEmits
+  const emit = defineEmits(["update:modelValue"]);
+
+  // when someVar change, it will update the reference passed via v-model
+  watch(someVar, (value) => {
+    emit("update:modelValue", value);
+  });
+</script>
+```
+
+## Make use of the modal
+
+To be able to control and populate data into the modal, we'll use the store defined as follow:
+
+```html
+<template>
+  <div>
+    <button class="btn" @click="handleOnClickOpenModal">Open</button>
+  </div>
+</template>
+
+<script lang="ts" setup>
+  import useModal from "@/stores/modal";
+  import MyViewToRender from "@/views/MyViewToRender.vue";
+
+  const modal = useModal();
+  function handleOnClickOpenModal() {
+    modal.open(MyViewToRender, [
+      {
+        label: "Save",
+        callback: (dataFromView) => {
+          console.log(dataFromView);
+          modal.close();
+        },
+      },
+    ]);
+  }
+</script>
+```
+
+When a click event on the `button` occurs, it will call `handleOnClickOpenModal` which will call the action `open` from the store, which changes the `isOpen` variable from the state, and as the modal is observing this value it will render or not depending of the value as follow:
+
+🎉 Hurra!!! you have made a reusable modal on top of composition API and pinia.
+
+![Dynamic modal on vue 3](./assets/reusable-modal-on-vue3.gif)
+
+### What if we want to pass a payload to our view?
+
+_Dependency injection is a programming technique / design pattern in which an object or function receives other objects or functions that it depends on._
+
+Vue natively implements a mechanism to handle this pattern, this implementation is known as `Provide / Inject`, more info [here](https://vuejs.org/guide/components/provide-inject.html). A parent component can serve as a dependency provider for all its descendants. Any component in the descendant tree, regardless of how deep it is, can inject dependencies provided by components up in its parent chain. The issue with this approach is it only applies to direct descendants from parent, not parallel components and as the modal view is not strictly a child of the component invoking it, it wont work as intended.
```

Comment:
```suggestion
Vue natively implements a mechanism to handle this pattern, this implementation is known as `Provide / Inject`, more info [here](https://vuejs.org/guide/components/provide-inject.html). A parent component can serve as a dependency provider for all its descendants. Any component in the descendant tree, regardless of how deep it is, can inject dependencies provided by components up in its parent chain. The issue with this approach is that it only applies to direct descendants from parent (not parallel components) and as the modal view is not strictly a child of the component invoking it, it won't work as intended.
```

Reply:
al `how deep it is` le falta un participio / adjetivo 🤔

Reply:
Good catch, gracias!